File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public static boolean cando(int capacity,int[]weights,int days){
        int c=0;
        int d=1;
        int i=0;
        for( i=0;i<weights.length;i++){
            if((c+weights[i])<=capacity){
                c+=weights[i];
            }else{
                d++;
                c=0;
                if(weights[i]<=capacity){
                     c+=weights[i];

                }else{
                    break;
                }
               
            }
            }
            if(i!=weights.length){
                return false;
            }
            if(d>days){
                return false;
            }
            return true;
    
    }
    public int shipWithinDays(int[] weights, int days) {
       
        int maxCapacity=0;
        for(int i=0;i<weights.length;i++){
               maxCapacity+=weights[i];
        }
        int left=0;
        int right=maxCapacity;
        while(left<right){
            int mid=(left+right)/2;
            if(cando(mid,weights,days))
            {
                right=mid;
            }else{
                left=mid+1;
            }
        }
        return left;
        
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
bool canbePossible(vector<int>&weights,int mid,int days){
    int sum=0;
    int ans=1;
    for(int i=0;i<weights.size();i++){
        sum+=weights[i];
        if(sum>mid){
            sum=weights[i];
            ans+=1;
            if(ans>days){
                return false;
            }
        }

    }
    return true;
}
    int shipWithinDays(vector<int>& weights, int days) {
        int left=*max_element(weights.begin(),weights.end());
        int right=0;
        for(auto it:weights){
            right+=it;
        }
        while(left<right){
            int mid=left+(right-left)/2;
            if(canbePossible(weights,mid,days)){
                right=mid;
            }else{
                left=mid+1;
            }
        }
        return left;

        
    }
};